File: address-geocode-us-python/REST/get_best_match_v4_rest.py
import requests
from agus_response import GetBestMatchV4Response, Error, InformationComponent

# Endpoint URLs for ServiceObjects Address Geocode US (AGUS) API
primary_url = "https://sws.serviceobjects.com/GCR/api.svc/json/GetBestMatch_V4?"
backup_url = "https://swsbackup.serviceobjects.com/GCR/api.svc/json/GetBestMatch_V4?"
trial_url = "https://trial.serviceobjects.com/GCR/api.svc/json/GetBestMatch_V4?"

def get_best_match_v4(
    address: str,
    city: str,
    state: str,
    postal_code: str,
    license_key: str,
    is_live: bool = True) -> GetBestMatchV4Response:
    """
    Call ServiceObjects Address Geocode US (AGUS) API's GetBestMatch_V4 endpoint
    to retrieve geocoding information (latitude, longitude, ZIP code, etc.).

    Parameters:
        address: Address line of the address to geocode (e.g., "123 Main Street").
        city: The city of the address to geocode. For example, "New York".
        state: The state of the address to geocode. For example, "NY".
        postal_code: The postal code of the address to geocode. For example, "10001".
        license_key: Your ServiceObjects license key.
        is_live: Use live or trial servers.

    Returns:
        GetBestMatchV4Response: Parsed JSON response with geocoding results or error details.

    Raises:
        RuntimeError: If the API call fails on both primary and backup endpoints or if JSON parsing fails.
    """
    params = {
        "Address": address,
        "City": city,
        "State": state,
        "PostalCode": postal_code,
        "LicenseKey": license_key,
    }
    # Select the base URL: production vs trial
    url = primary_url if is_live else trial_url

    try:
        # Attempt primary (or trial) endpoint
        response = requests.get(url, params=params, timeout=10)
        response.raise_for_status()
        data = response.json()

         # If API returned an error in JSON payload, trigger fallback
        error = getattr(response, 'Error', None)
        if not (error is None or getattr(error, 'Number', None) != "4"):
            if is_live:
                # Try backup URL
                response = requests.get(backup_url, params=params, timeout=10)
                data = response.json()

                # If still error, propagate exception
                if 'Error' in data:
                    raise RuntimeError(f"AGUS service error: {data['Error']}")
            else:
                # Trial mode error is terminal
                raise RuntimeError(f"AGUS trial error: {data['Error']}")

        # Convert JSON response to GetBestMatchV4Response for structured access
        error = Error(**data.get("Error", {})) if data.get("Error") else None

        return GetBestMatchV4Response(
          Level=data.get("Level"),
          LevelDescription=data.get("LevelDescription"),
          Latitude=data.get("Latitude"),
          Longitude=data.get("Longitude"),
          Zip=data.get("Zip"),
          InformationComponents=[
              InformationComponent(Name=comp.get("Name"), Value=comp.get("Value"))
              for comp in data.get("InformationComponents", [])
          ] if "InformationComponents" in data else [],
            Error=error,
          )

    except requests.RequestException as req_exc:
        # Network or HTTP-level error occurred
        if is_live:
            try:
                # Fallback to backup URL
                response = requests.get(backup_url, params=params, timeout=10)
                response.raise_for_status()
                data = response.json()
                if "Error" in data:
                    raise RuntimeError(f"AGUS backup error: {data['Error']}") from req_exc

                error = Error(**data.get("Error", {})) if data.get("Error") else None

                return GetBestMatchV4Response(
                    Level=data.get("Level"),
                    LevelDescription=data.get("LevelDescription"),
                    Latitude=data.get("Latitude"),
                    Longitude=data.get("Longitude"),
                    Zip=data.get("Zip"),
                    InformationComponents=[
                        InformationComponent(Name=comp.get("Name"), Value=comp.get("Value"))
                        for comp in data.get("InformationComponents", [])
                    ] if "InformationComponents" in data else [],
                    Error=error,
                )
            except Exception as backup_exc:
                raise RuntimeError("AGUS service unreachable on both endpoints") from backup_exc
        else:
            raise RuntimeError(f"AGUS trial error: {str(req_exc)}") from req_exc

File: address-geocode-us-nodejs/REST/get_best_match_v4_rest.js
import axios from 'axios';
import querystring from 'querystring';
import { GetBestMatchV4Response } from './agus_response.js';

/**
 * @constant
 * @type {string}
 * @description The base URL for the live ServiceObjects Address Geocode US (AGUS) API service.
 */
const liveBaseUrl = 'https://sws.serviceobjects.com/GCR/api.svc/json/';

/**
 * @constant
 * @type {string}
 * @description The base URL for the backup ServiceObjects Address Geocode US (AGUS) API service.
 */
const backupBaseUrl = 'https://swsbackup.serviceobjects.com/GCR/api.svc/json/';

/**
 * @constant
 * @type {string}
 * @description The base URL for the trial ServiceObjects Address Geocode US (AGUS) API service.
 */
const trialBaseUrl = 'https://trial.serviceobjects.com/GCR/api.svc/json/';

/**
 * <summary>
 * Checks if a response from the API is valid by verifying that it either has no Error object
 * or the Error.Number is not equal to '4'.
 * </summary>
 * <param name="response" type="Object">The API response object to validate.</param>
 * <returns type="boolean">True if the response is valid, false otherwise.</returns>
 */
const isValid = (response) => !response?.Error || response.Error.Number !== '4';

/**
 * <summary>
 * Constructs a full URL for the GetBestMatch_V4 API endpoint by combining the base URL
 * with query parameters derived from the input parameters.
 * </summary>
 * <param name="params" type="Object">An object containing all the input parameters.</param>
 * <param name="baseUrl" type="string">The base URL for the API service (live, backup, or trial).</param>
 * <returns type="string">The constructed URL with query parameters.</returns>
 */
const buildUrl = (params, baseUrl) =>
    `${baseUrl}GetBestMatch_V4?${querystring.stringify(params)}`;

/**
 * <summary>
 * Performs an HTTP GET request to the specified URL with a given timeout.
 * </summary>
 * <param name="url" type="string">The URL to send the GET request to.</param>
 * <param name="timeoutSeconds" type="number">The timeout duration in seconds for the request.</param>
 * <returns type="Promise<GetBestMatchV4Response>">A promise that resolves to a GetBestMatchV4Response object containing the API response data.</returns>
 * <exception cref="Error">Thrown if the HTTP request fails, with a message detailing the error.</exception>
 */
const httpGet = async (url, timeoutSeconds) => {
    try {
        const response = await axios.get(url, { timeout: timeoutSeconds * 1000 });
        return new GetBestMatchV4Response(response.data);
    } catch (error) {
        throw new Error(`HTTP request failed: ${error.message}`);
    }
};

/**
 * <summary>
 * Provides functionality to call the ServiceObjects Address Geocode US (AGUS) API's GetBestMatch_V4 endpoint,
 * retrieving geocoding information (e.g., latitude, longitude, ZIP code) for a given US address with fallback to a backup endpoint for reliability in live mode.
 * </summary>
 */
const GetBestMatchV4Client = {
    /**
     * <summary>
     * Asynchronously invokes the GetBestMatch_V4 API endpoint, attempting the primary endpoint
     * first and falling back to the backup if the response is invalid (Error.Number == '4') in live mode.
     * </summary>
     * @param {string} Address - Address line of the address to geocode (e.g., "123 Main Street").
     * @param {string} City - The city of the address to geocode. For example, “New York”. The city isn’t required, but if one is not provided, the Zip code is required..
     * @param {string} State - The state of the address to geocode. For example, “NY”. This does not need to be contracted, full state names will work as well. The state isn’t required, but if one is not provided, the Zip code is required..
     * @param {string} PostalCode - The state of the address to geocode. For example, “NY”. This does not need to be contracted, full state names will work as well. The state isn’t required, but if one is not provided, the Zip code is required..
     * @param {string} LicenseKey - Your license key to use the service.
     * @param {boolean} isLive - Value to determine whether to use the live or trial servers.
     * @param {number} timeoutSeconds - Timeout, in seconds, for the call to the service.
     * @returns {Promise<GetBestMatchV4Response>} - A promise that resolves to a GetBestMatchV4Response object.
     */
    async invokeAsync(Address, City, State, PostalCode, LicenseKey, isLive = true, timeoutSeconds = 15) {
        const params = {
            Address,
            City,
            State,
            PostalCode,
            LicenseKey
        };

        const url = buildUrl(params, isLive ? liveBaseUrl : trialBaseUrl);
        let response = await httpGet(url, timeoutSeconds);

        if (isLive && !isValid(response)) {
            const fallbackUrl = buildUrl(params, backupBaseUrl);
            const fallbackResponse = await httpGet(fallbackUrl, timeoutSeconds);
            return fallbackResponse;
        }
        return response;
    },

    /**
     * <summary>
     * Synchronously invokes the GetBestMatch_V4 API endpoint by wrapping the async call
     * and awaiting its result immediately.
     * </summary>
     * @param {string} Address - Address line of the address to geocode (e.g., "123 Main Street").
     * @param {string} City - The city of the address to geocode. For example, “New York”. The city isn’t required, but if one is not provided, the Zip code is required..
     * @param {string} State - The state of the address to geocode. For example, “NY”. This does not need to be contracted, full state names will work as well. The state isn’t required, but if one is not provided, the Zip code is required..
     * @param {string} PostalCode - The state of the address to geocode. For example, “NY”. This does not need to be contracted, full state names will work as well. The state isn’t required, but if one is not provided, the Zip code is required..
     * @param {string} LicenseKey - Your license key to use the service.
     * @param {boolean} isLive - Value to determine whether to use the live or trial servers.
     * @param {number} timeoutSeconds - Timeout, in seconds, for the call to the service.
     * @returns {GetBestMatchV4Response} - A GetBestMatchV4Response object with geocoding details or an error.
     */
    invoke(Address, City, State, PostalCode, LicenseKey, isLive = true, timeoutSeconds = 15) {
        return (async () => await this.invokeAsync(
            Address, City, State, PostalCode, LicenseKey, isLive, timeoutSeconds
        ))();
    }
};

export { GetBestMatchV4Client, GetBestMatchV4Response };